R/callback.R

```diff
@@ -0,0 +1,67 @@
+as_chunk_callback <- function(x) UseMethod("as_chunk_callback")
+as_chunk_callback.function <- function(x) {
+  SideEffectChunkCallback$new(x)
+}
+as_chunk_callback.R6ClassGenerator <- function(x) {
+  as_chunk_callback(x$new())
+}
+as_chunk_callback.ChunkCallback <- function(x) {
+  x
+}
+
+ChunkCallback <- R6::R6Class("ChunkCallback",
+  private = list(
+    callback = NULL
+  ),
+  public = list(
+    initialize = function(callback) NULL,
+    receive = function(data, index) NULL,
+    continue = function() TRUE,
+    result = function() NULL,
+    finally = function() NULL
+  )
+)
+
+# This would be used if the result should be thrown away
+SideEffectChunkCallback <- R6::R6Class("SideEffectChunkCallback", inherit = ChunkCallback,
+  private = list(
+     cancel = FALSE
+    ),
+  public = list(
+    initialize = function(callback) {
+      check_callback_fun(callback)
+      private$callback <- callback
+    },
+    receive = function(data, index) {
+      result <- private$callback(data, index)
+      private$cancel <- identical(result, FALSE)
+    },
+    continue = function() {
+      !private$cancel
+    }
+  )
+)
+
+# Used if the result of each chunk should be combined
+# at the end
+DataFrameCallback <- R6::R6Class("DataFrameCallback", inherit = ChunkCallback,
+  public = list(
+    initialize = function(callback) {
+      private$callback <- callback
+    },
+    receive = function(data, index) {
+      result <- private$callback(data, index)
+      private$results <- c(private$results, list(result))
+    },
+    result = function() {
+      dplyr::bind_rows(private$results)
```

Comment:
I went with rbind for now, I think it is a useful example class, we can always have a more performant/robust one in the future package.

The DataFrameCallback needed a slight tweak to work anyway, you need to define the retrieve variable or else you get a locked namespace error when R tries to create the new variable in the object's environment.